nexus/types/src/deployment/tri_map.rs

```diff
@@ -0,0 +1,509 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::{
+    borrow::Borrow,
+    collections::{hash_map, BTreeSet, HashMap},
+    fmt,
+    hash::Hash,
+};
+
+use derive_where::derive_where;
+use serde::{Deserialize, Serialize, Serializer};
+
+/// An append-only 1:1:1 (trijective) map for three keys and a value.
+///
+/// The storage mechanism is a vector of entries, with indexes into that vector
+/// stored in three hashmaps. This allows for efficient lookups by any of the
+/// three keys, while preventing duplicates.
+///
+/// Not totally generic yet, just meant for the deployment use case.
+#[derive_where(Clone, Debug, Default)]
+pub(crate) struct TriMap<T: TriMapEntry> {
+    entries: Vec<T>,
+    // Invariant: the value (usize) in these maps are valid indexes into
+    // `entries`, and are a 1:1 mapping.
+    k1_to_entry: HashMap<T::K1, usize>,
+    k2_to_entry: HashMap<T::K2, usize>,
+    k3_to_entry: HashMap<T::K3, usize>,
+}
+
+// Note: Eq and PartialEq are not implemented for TriMap. Implementing them
+// would need to be done with care, because
```

Comment:
This comment cuts off just when it's getting good. :)

Reply:
Oof, meant to write something like:

```
// Note: Eq and PartialEq are not implemented for TriMap. Implementing them
// would need to be done with care, because TriMap is not semantically like an
// IndexMap: two maps are equivalent even if their entries are in a different
// order.
```

Reply:
Fixed.